### ISSUE TITLE
Create hooks and ops for adding and removing Constraints and Springs

### ISSUE THREAD

**Issue by nevernotsean** [react-spring/use-cannon#14]
Connecting two rigidbodies together can be done with Constraints and Springs. If I understand CANNONjs correctly, Constraints are rather rigid (a set distance apart) while Springs can stretch and expand with forces.

Currently the worker can add and remove bodies as well as initiate the world, what we should add is addConstraint, removeConstraint, addSpring, removeSpring operations which handle constraints in a similar way to bodies. Then, new useConstraint and useSpring hooks would accept two bodies as arguments (plus whatever options are needed).

This could lead to some really cool results such as physics driven Ragdolls, Rope/Cables/Hair, Cloth, Vehicle physics, etc.

**Comment by nevernotsean** [comment]
related to #6, this task would need to be done before

**Comment by nevernotsean** [comment]
@drcmda I'm going to dive into this, any thoughts on how we should go about connecting two bodies? My thought was to provide a hook two UUIDs and config arguments that creates a constraint/spring between the two.

**Comment by drcmda** [comment]
sure, that sounds fantastic. im learning physics watching you :-P

**Comment by codynova** [comment]
I think it would be cool if the `useSpring` hook allowed passing two Three.Object3D refs for succinctness. Something similar to this perhaps?

```js
const [ box ] = useBox(() => ({}))
const [ sphere ] = useSphere(() => ({}))
const [ spring ] = useSpring(box, sphere, {})
```

**Comment by drcmda** [comment]
i also think refs are probably a better way than internal uuids. and that would even work across components due to React.forwardRef

**Comment by drcmda** [comment]
btw, this is related to https://github.com/react-spring/use-cannon/issues/27

**Comment by nevernotsean** [comment]
Made a feature branch for all constraints, I've got plain constraints working with a example:
https://github.com/react-spring/use-cannon/tree/feature/constraints

**Comment by codynova** [comment]
This is great!

Can we change the hook signature to accept the two bodies before the offsets, and make `maxForce` optional? The internal Cannon default for maxForce is `1e6`. This would allow defining a PointToPointConstraint with only the two body refs.

```
export function useConstraint(
  bodyA: React.MutableRefObject<THREE.Object3D | undefined>,
  bodyB: React.MutableRefObject<THREE.Object3D | undefined>,
  offsetA: [number, number, number] = [0, 0, 0],
  offsetB: [number, number, number] = [0, 0, 0],
  maxForce?: number
)
```

I'm assuming we can add the additional constraint types into the `addConstraint` worker op?

**Comment by codynova** [comment]
I think we could update the `useConstraint` hook to accept a constraint type string and options object (according to Cannon api [base Constraint class](https://schteppe.github.io/cannon.js/docs/classes/Constraint.html)) and return a unique constraint api - different from the body api. Then we could expose methods on that api to enable/disable the constraint, etc. Then we could have additional hooks for each unique constraint type that call `useConstraint` themselves - similar to the way `useBody` is consumed by the specific shape hooks.

What do you think @drcmda

**Comment by nevernotsean** [comment]
Yes, that makes sense. Also when checking out the docs of the other types of constraints, the api is rather inconsistent. I'm thinking of doing this to keep the first two parameters just bodies, then an option object. Trying to use options that don't apply would just be ignored:

`type ConstraintOptionTypes = {
  type: 'PointToPoint' | 'ConeTwist' | 'Distance' | 'Hinge' | 'Lock'
  pivotA: [number, number, number]
  pivotB: [number, number, number]
  axisA: [number, number, number]
  axisB: [number, number, number]
  maxForce: number
  collideConnected: boolean
  wakeUpBodies: boolean
}

export function useConstraint(
  bodyA: React.MutableRefObject<THREE.Object3D | undefined>,
  bodyB: React.MutableRefObject<THREE.Object3D | undefined>,
  options: ConstraintOptionTypes
)`

**Comment by nevernotsean** [comment]
Also, I agree on your second comment @codynova. Using the way `useBody` types are implemented would keep everything nice and typed.

**Comment by codynova** [comment]
Paul may have more recommendations around API structure - he has a lot of experience with developer experience of more complicated APIs like react-spring, and I wouldn't say it's my strong suit 😄 

Great work so far! I hope to have something to show for compound shapes tonight

**Comment by codynova** [comment]
I also think that improving the typings of Cannon, and adding discriminated union types to `use-cannon` where appropriate, would clear up a lot of the confusion around when to use specific props/args - especially if the `use-cannon` user isn't familiar with the Cannon core API at all

**Comment by nevernotsean** [comment]
True, to be honest this is my first try using typescript, so I'll let you and @drcmda take the lead on that. Coming from c# its mostly intuitive, definitely seeing the benefits already!

**Comment by nevernotsean** [comment]
Constraints and Spring added! I created a PR so we can discuss if needed #30

**Comment by codynova** [comment]
Awesome! I was only able to look briefly, but it's looking good. I left a few small comments and added Paul as a reviewer. 👀 

I noticed in the Cannon docs the base Spring class allows enabling/disabling, so we may want to return a constraint api from `useConstraint` in the future, but I think it's a good idea to make sure this foundation is as solid as it can be first! 😸

**Comment by codynova** [comment]
@nevernotsean This is all finished, right? From what I can see in the hooks it's looking pretty complete

**Comment by nevernotsean** [comment]
Thanks, yep this is done, if there are any enhancements to Constraints/Springs feature we can create a separate issue. closing!
